### PR TITLE
Clean up multiservice-discovery to be more robust, lock shared struct…

### PR DESCRIPTION
**Pull request by DFINITYManu**
…ures, and separate concerns.

In particular, the use of parallel lists for thread end signaling channels and actual threads has been eliminated.

Additional changes:

* improved logging,
* improved messages issued to the user via HTTP,
* graceful interruption of initial syncs,
* parallel stop and start of multiple definitions submitted via POST / PUT,
* typing improvements (better error typing, structural typing to prevent usage bugs with the DefinitionSupervisor).

Preliminary work to ensure other modifications are possible.

### CODE REVIEW

**Review by NikolaMilosa — APPROVED** [review]
Nice! I like the distinction between running and non-running definitions. The only thing I am not a fan of are the comments. While they do help sometimes I feel like a lot of comments don't allow me to really understand the code but actually push me towards thinking that is what the code does without reviewing it. I am okay with some comments before the function which explains what the function on high level does but if we need comments in the flow I think we are just having bad names. If we added these comments above the functions we could add some sort of exporter which would generate docs for us and include it in our github pages in the future.

File: rs/ic-observability/multiservice-discovery/src/server_handlers/add_boundary_node_to_definition_handler.rs [hunk 1]
```diff
@@ -1,79 +1,86 @@
...
     pub log: Logger,
 }
 
-pub async fn add_boundary_node(
+#[derive(Debug)]
+pub enum AddBoundaryNodeError {
+    DefinitionNotFound(String),
+    BoundaryNodeAlreadyExists(BoundaryNodeAlreadyExists),
+    BadBoundaryNodeDtoError(BadBoundaryNodeDtoError),
+}
```
[the remark this answers is not in the document]
Comment (DFINITYManu) [reply rs/ic-observability/multiservice-discovery/src/server_handlers/add_boundary_node_to_definition_handler.rs:23]: I'm thinking the errors in these functions are now superfluous now that the management layer has moved down a level into the Supervisor.  They add a lot of code for little gain.

File: rs/ic-observability/multiservice-discovery/src/server_handlers/delete_definition_handler.rs [hunk 2]
```diff
@@ -1,37 +1,22 @@
...
-        None => Ok(warp::reply::with_status(
-            "Definition with this name does not exist".to_string(),
+    match supervisor.stop(vec![name.clone()]).await {
+        Ok(_) => Ok(warp::reply::with_status(
+            format!("Deleted definition {}", name.clone()),
+            warp::http::StatusCode::OK,
+        )),
+        Err(e) => Ok(warp::reply::with_status(
+            format!("Could not delete definition {}: {}", name, e),
             warp::http::StatusCode::BAD_REQUEST,
```
[the remark this answers is not in the document]
Comment (DFINITYManu) [reply rs/ic-observability/multiservice-discovery/src/server_handlers/delete_definition_handler.rs]: Multiple DELETE parameters makes no sense from a REST perspective.

If you want to delete multiple, the REST way is to `GET` the definitions, `sed` out those who don't need to be there anymore, then `PUT` again.

File: rs/ic-observability/multiservice-discovery/src/server_handlers/add_boundary_node_to_definition_handler.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (DFINITYManu) [reply rs/ic-observability/multiservice-discovery/src/server_handlers/add_boundary_node_to_definition_handler.rs]: I deleted a ton of error code that was redundant.  We'd delete them anyway when porting to AXUM.

**Comment by DFINITYManu** [comment]
I can't fix the broken test.  The test data is in protobuf format, and I can't tell what the structure is, so I can't tell what the output is supposed to look like.

**Comment by DFINITYManu** [comment]
```
     Running tests/tests.rs (target/debug/deps/tests-b845cb7b0ebbef8d)

running 1 test
test tests::prom_targets_tests ... FAILED

failures:

---- tests::prom_targets_tests stdout ----
thread 'tests::prom_targets_tests' panicked at rs/ic-observability/multiservice-discovery/tests/tests.rs:52:52:
called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    tests::prom_targets_tests

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

Can we delete this test and make a proper unit test instead?

**Comment by NikolaMilosa** [comment]
I believe this was supposed to be an integration test not a unit test. cc @pietrodimarco-dfinity

**Pull request merged**